pkg/trace/writer/trace.go

```diff
@@ -127,183 +122,199 @@ func NewTraceWriter(
+func (w *TraceWriter) Stop() {
+	log.Debug("Exiting trace writer. Trying to flush whatever is left...")
+	close(w.stop)
+	// Wait for encoding/compression to complete on each payload,
+	// and submission to senders
+	w.wg.Wait()
+	w.flush()
+	stopSenders(w.senders)
+}
+
 // FlushSync blocks and sends pending payloads when syncMode is true
 func (w *TraceWriter) FlushSync() error {
 	if !w.syncMode {
 		return errors.New("not flushing; sync mode not enabled")
 	}
 	defer w.report()
 
-	notify := make(chan struct{}, 1)
-	w.flushChan <- notify
-	<-notify
+	w.flush()
 	return nil
 }
 
-func (w *TraceWriter) addSpans(pkg *SampledChunks) {
-	w.stats.Spans.Add(pkg.SpanCount)
-	w.stats.Traces.Add(int64(len(pkg.TracerPayload.Chunks)))
-	w.stats.Events.Add(pkg.EventCount)
-
+// appendChunks adds sampled chunks to the current payload, and in the case the payload
+// is full, returns a finished payload which needs to be written out.
+func (w *TraceWriter) appendChunks(pkg *SampledChunks) []*pb.TracerPayload {
+	var toflush []*pb.TracerPayload
+	w.mu.Lock()
+	defer w.mu.Unlock()
 	size := pkg.Size
 	if size+w.bufferedSize > MaxPayloadSize {
 		// reached maximum allowed buffered size
-		w.flush()
+		// reset the buffer so we can add our payload and defer a flush.
+		toflush = w.tracerPayloads
+		w.resetBuffer()
 	}
 	if len(pkg.TracerPayload.Chunks) > 0 {
 		log.Tracef("Writer: handling new tracer payload with %d spans: %v", pkg.SpanCount, pkg.TracerPayload)
 		w.tracerPayloads = append(w.tracerPayloads, pkg.TracerPayload)
 	}
 	w.bufferedSize += size
+	return toflush
 }
 
-func (w *TraceWriter) drainAndFlush() {
-outer:
-	for {
-		select {
-		case pkg := <-w.In:
-			w.addSpans(pkg)
-		default:
-			break outer
-		}
+// WriteChunks serializes the provided chunks, enqueueing them to be sent
+func (w *TraceWriter) WriteChunks(pkg *SampledChunks) {
+	w.stats.Spans.Add(pkg.SpanCount)
+	w.stats.Traces.Add(int64(len(pkg.TracerPayload.Chunks)))
+	w.stats.Events.Add(pkg.EventCount)
+
+	toflush := w.appendChunks(pkg)
+	if toflush != nil {
+		w.flushPayloads(toflush)
 	}
-	w.flush()
-	w.swg.Wait()
 }
-
 func (w *TraceWriter) resetBuffer() {
 	w.bufferedSize = 0
 	w.tracerPayloads = make([]*pb.TracerPayload, 0, len(w.tracerPayloads))
 }
 
 const headerLanguages = "X-Datadog-Reported-Languages"
 
+// w must be locked for a flush.
 func (w *TraceWriter) flush() {
-	if len(w.tracerPayloads) == 0 {
+	w.mu.Lock()
+	defer w.mu.Unlock()
+	defer w.resetBuffer()
+	w.flushPayloads(w.tracerPayloads)
+}
+
+// This pool is used to save GC pressure from having to re-allocate gzip writers for every compress operation.
+var gzpool = sync.Pool{}
+
+func getGZW(w io.Writer) (*gzip.Writer, error) {
+	gw := gzpool.Get()
+	if gw == nil {
+		return gzip.NewWriterLevel(w, gzip.BestSpeed)
+	}
+	gws := gw.(*gzip.Writer)
+	gws.Reset(w)
+	return gws, nil
+}
+
+// w does not need to be locked during flushPayloads.
```

Comment:
This will have conflicts with zstd PR.

Reply:
Yep :P I'm expecting to have to modify this (or that PR) to support this perf change